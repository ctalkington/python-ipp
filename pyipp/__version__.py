"""Asynchronous Python client for IPP."""

__version__ = "0.11.1"
