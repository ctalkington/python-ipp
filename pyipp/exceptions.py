"""Exceptions for IPP."""


class IPPError(Exception):
    """Generic IPP exception."""

    pass


class IPPConnectionError(IPPError):
    """IPP connection exception."""

    pass


class IPPConnectionUpgradeRequired(IPPError):
    """IPP connection upgrade requested."""

    pass


class IPPParseError(IPPError):
    """IPP parse exception."""

    pass


class IPPResponseError(IPPError):
    """IPP response exception."""

    pass
