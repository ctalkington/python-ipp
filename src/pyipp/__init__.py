"""Asynchronous Python client for IPP."""
from .ipp import (  # noqa
    IPP,
    IPPConnectionError,
    IPPConnectionUpgradeRequired,
    IPPError,
    IPPParseError,
    IPPResponseError,
    IPPVersionNotSupportedError,
)
from .models import Printer  # noqa
