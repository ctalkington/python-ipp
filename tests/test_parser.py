"""Tests for Parser."""
import pytest
from syrupy.assertion import SnapshotAssertion

from pyipp import IPPParseError, parser
from pyipp.const import DEFAULT_CHARSET, DEFAULT_CHARSET_LANGUAGE, DEFAULT_PROTO_VERSION
from pyipp.enums import IppOperation

from . import load_fixture_binary

RESPONSE_GET_PRINTER_ATTRIBUTES = load_fixture_binary(
    "get-printer-attributes-response-000.bin",
)

MOCK_IEEE1284_DEVICE_ID = "MFG:EPSON;CMD:ESCPL2,BDC,D4,D4PX,ESCPR7,END4,GENEP,URF;MDL:XP-6000 Series;CLS:PRINTER;DES:EPSON XP-6000 Series;CID:EpsonRGB;FID:FXN,DPA,WFA,ETN,AFN,DAN,WRA;RID:20;DDS:022500;ELG:1000;SN:583434593035343012;URF:CP1,PQ4-5,OB9,OFU0,RS360,SRGB24,W8,DM3,IS1-7-6,V1.4,MT1-3-7-8-10-11-12;"


def test_parse() -> None:
    """Test the parse method."""
    result = parser.parse(RESPONSE_GET_PRINTER_ATTRIBUTES)
    assert result == {
        "data": b"",
        "jobs": [],
        "operation-attributes": {
            "attributes-charset": DEFAULT_CHARSET,
            "attributes-natural-language": DEFAULT_CHARSET_LANGUAGE,
            "printer-uri": "ipp://printer.example.com:361/ipp/print",
            "requesting-user-name": "PythonIPP",
        },
        "printers": [],
        "request-id": 1,
        "status-code": IppOperation.GET_PRINTER_ATTRIBUTES,
        "unsupported-attributes": [],
        "version": DEFAULT_PROTO_VERSION,
    }


def test_parse_attribute() -> None:
    """Test the parse_attribute method."""
    result = parser.parse_attribute(RESPONSE_GET_PRINTER_ATTRIBUTES, 9)
    assert result == (
        {
            "name": "attributes-charset",
            "name-length": 18,
            "tag": 71,
            "value": "utf-8",
            "value-length": 5,
        },
        37,
    )


def test_parse_attribute_reserved_string() -> None:
    """Test the parse_attribute method when provided a reserved string."""
    result = parser.parse_attribute(b"C\x00\x0freserved-string\x00\x04yoda", 0)
    assert result == (
        {
            "name": "reserved-string",
            "name-length": 15,
            "tag": 67,
            "value": "yoda",
            "value-length": 4,
        },
        24,
    )

    result = parser.parse_attribute(b"C\x00\x0freserved-string\x00\x00", 0)
    assert result == (
        {
            "name": "reserved-string",
            "name-length": 15,
            "tag": 67,
            "value": None,
            "value-length": 0,
        },
        20,
    )


def test_parse_attribute_invalid_date() -> None:
    """Test the parse_attribute method when provided an invalid date."""
    invalid = b"1\x00\x14printer-current-time\x00\x0299"

    with pytest.raises(IPPParseError):
        parser.parse_attribute(invalid, 0)


def test_parse_ieee1284_device_id() -> None:
    """Test the parse_ieee1284_device_id method."""
    result = parser.parse_ieee1284_device_id(MOCK_IEEE1284_DEVICE_ID)

    assert result
    assert result["MFG"] == "EPSON"
    assert result["MDL"] == "XP-6000 Series"
    assert result["SN"] == "583434593035343012"
    assert result["CMD"] == "ESCPL2,BDC,D4,D4PX,ESCPR7,END4,GENEP,URF"

    assert result["MANUFACTURER"] == result["MFG"]
    assert result["MODEL"] == result["MDL"]
    assert result["COMMAND SET"] == result["CMD"]


def test_parse_ieee1284_device_id_manufacturer_only() -> None:
    """Test the parse_ieee1284_device_id method with only a manufacturer."""
    result = parser.parse_ieee1284_device_id("MANUFACTURER:EPSON")

    assert result == {
        "MANUFACTURER": "EPSON",
    }


def test_parse_ieee1284_device_id_empty() -> None:
    """Test the parse_ieee1284_device_id method with empty string."""
    result = parser.parse_ieee1284_device_id("")

    assert isinstance(result, dict)


def test_parse_make_and_model() -> None:
    """Test the parse_make_and_model method."""
    result = parser.parse_make_and_model("")
    assert result == ("Unknown", "Unknown")

    # generic fallback for unknown brands
    result = parser.parse_make_and_model("IPP")
    assert result == ("IPP", "Unknown")

    result = parser.parse_make_and_model("IPP Printer")
    assert result == ("IPP", "Printer")

    # known brands
    result = parser.parse_make_and_model("EPSON XP-6000 Series")
    assert result == ("EPSON", "XP-6000 Series")

    result = parser.parse_make_and_model("HP Officejet Pro 6830")
    assert result == ("HP", "Officejet Pro 6830")

    result = parser.parse_make_and_model("HP Photosmart D110 Series")
    assert result == ("HP", "Photosmart D110 Series")


def test_parse_brother_mfcj5320dw(snapshot: SnapshotAssertion) -> None:
    """Test the parse method against response from Brother MFC-J5320DW."""
    response = load_fixture_binary("get-printer-attributes-brother-mfcj5320dw.bin")

    result = parser.parse(response)
    assert result == snapshot


def test_parse_epson_xp6000(snapshot: SnapshotAssertion) -> None:
    """Test the parse method against response from Epson XP-6000 Series."""
    response = load_fixture_binary("get-printer-attributes-epsonxp6000.bin")

    result = parser.parse(response)
    assert result == snapshot


def test_parse_kyocera_ecosys_m2540dn(snapshot: SnapshotAssertion) -> None:
    """Test the parse method against response from Kyocera Ecosys M2540DN."""
    response = load_fixture_binary(
        "get-printer-attributes-kyocera-ecosys-m2540dn-001.bin",
    )

    result = parser.parse(response)
    assert result == snapshot
