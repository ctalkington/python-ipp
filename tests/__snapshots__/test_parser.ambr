# serializer version: 1
# name: test_parse_brother_mfcj5320dw
  dict({
    'data': b'',
    'jobs': list([
    ]),
    'operation-attributes': dict({
      'attributes-charset': 'utf-8',
      'attributes-natural-language': 'de',
    }),
    'printers': list([
      dict({
        'charset-configured': 'utf-8',
        'charset-supported': 'utf-8',
        'color-supported': True,
        'compression-supported': 'none',
        'copies-default': 1,
        'copies-supported': list([
          1,
          99,
        ]),
        'document-format-default': 'application/octet-stream',
        'document-format-supported': list([
          'application/octet-stream',
          'application/vnd.brother-hbp',
          'image/pwg-raster',
          'image/urf',
          'image/jpeg',
        ]),
        'finishings-default': <IppFinishing.NONE: 3>,
        'finishings-supported': <IppFinishing.NONE: 3>,
        'generated-natural-language-supported': 'de',
        'identify-actions-default': list([
          'flash',
          'sound',
        ]),
        'identify-actions-supported': list([
          'flash',
          'sound',
        ]),
        'ipp-features-supported': list([
          'airprint-1.3',
          'faxout',
          'wfds-print-1.0',
        ]),
        'ipp-versions-supported': list([
          '1.0',
          '1.1',
          '2.0',
        ]),
        'job-creation-attributes-supported': list([
          'copies',
          'finishings',
          'ipp-attribute-fidelity',
          'job-name',
          'media',
          'media-col',
          'orientation-requested',
          'output-bin',
          'output-mode',
          'print-quality',
          'printer-resolution',
          'requesting-user-name',
          'sides',
          'print-color-mode',
        ]),
        'job-impressions-supported': list([
          1,
          999,
        ]),
        'jpeg-k-octets-supported': list([
          0,
          12288,
        ]),
        'jpeg-x-dimension-supported': list([
          0,
          8192,
        ]),
        'jpeg-y-dimension-supported': list([
          1,
          20000,
        ]),
        'landscape-orientation-requested-preferred': 5,
        'marker-colors': list([
          '#FF00FF',
          '#00FFFF',
          '#FFFF00',
          '#000000',
        ]),
        'marker-high-levels': list([
          100,
          100,
          100,
          100,
        ]),
        'marker-levels': list([
          11,
          9,
          45,
          11,
        ]),
        'marker-low-levels': list([
          18,
          18,
          18,
          18,
        ]),
        'marker-names': list([
          'M',
          'C',
          'Y',
          'BK',
        ]),
        'marker-types': list([
          'ink-cartridge',
          'ink-cartridge',
          'ink-cartridge',
          'ink-cartridge',
        ]),
        'media-bottom-margin-supported': list([
          300,
          0,
          2200,
          2200,
          1200,
        ]),
        'media-col-default': dict({
          'media-bottom-margin': 300,
          'media-left-margin': 300,
          'media-right-margin': 300,
          'media-size': dict({
            'x-dimension': 21000,
            'y-dimension': 29700,
          }),
          'media-source': 'main',
          'media-source-properties': dict({
            'media-source-feed-direction': 'long-edge-first',
            'media-source-feed-orientation': <IppOrientationRequested.REVERSE_LANDSCAPE: 5>,
          }),
          'media-top-margin': 300,
          'media-type': 'stationery',
        }),
        'media-col-ready': list([
          dict({
            'media-bottom-margin': 300,
            'media-left-margin': 300,
            'media-right-margin': 300,
            'media-size': dict({
              'x-dimension': 21000,
              'y-dimension': 29700,
            }),
            'media-source': 'main',
            'media-source-properties': dict({
              'media-source-feed-direction': 'long-edge-first',
              'media-source-feed-orientation': <IppOrientationRequested.REVERSE_LANDSCAPE: 5>,
            }),
            'media-top-margin': 300,
            'media-type': 'stationery',
          }),
          dict({
            'media-bottom-margin': 0,
            'media-left-margin': 0,
            'media-right-margin': 0,
            'media-size': dict({
              'x-dimension': 21000,
              'y-dimension': 29700,
            }),
            'media-source': 'main',
            'media-source-properties': dict({
              'media-source-feed-direction': 'long-edge-first',
              'media-source-feed-orientation': <IppOrientationRequested.REVERSE_LANDSCAPE: 5>,
            }),
            'media-top-margin': 0,
            'media-type': 'stationery',
          }),
        ]),
        'media-col-supported': list([
          'media-type',
          'media-size',
          'media-top-margin',
          'media-left-margin',
          'media-right-margin',
          'media-bottom-margin',
          'media-source',
        ]),
        'media-default': 'iso_a4_210x297mm',
        'media-left-margin-supported': list([
          300,
          0,
          300,
          300,
          300,
        ]),
        'media-ready': 'iso_a4_210x297mm',
        'media-right-margin-supported': list([
          300,
          0,
          300,
          300,
          300,
        ]),
        'media-size-supported': list([
          dict({
            'x-dimension': 21000,
            'y-dimension': 29700,
          }),
          dict({
            'x-dimension': 21590,
            'y-dimension': 27940,
          }),
          dict({
            'x-dimension': 18415,
            'y-dimension': 26670,
          }),
          dict({
            'x-dimension': 14800,
            'y-dimension': 21000,
          }),
          dict({
            'x-dimension': 10500,
            'y-dimension': 14800,
          }),
          dict({
            'x-dimension': 10160,
            'y-dimension': 15240,
          }),
          dict({
            'x-dimension': 12700,
            'y-dimension': 20320,
          }),
          dict({
            'x-dimension': 8890,
            'y-dimension': 12700,
          }),
          dict({
            'x-dimension': 12700,
            'y-dimension': 17780,
          }),
          dict({
            'x-dimension': 10477,
            'y-dimension': 24130,
          }),
          dict({
            'x-dimension': 11000,
            'y-dimension': 22000,
          }),
          dict({
            'x-dimension': 9842,
            'y-dimension': 19050,
          }),
          dict({
            'x-dimension': 21590,
            'y-dimension': 35560,
          }),
          dict({
            'x-dimension': 16200,
            'y-dimension': 22900,
          }),
          dict({
            'x-dimension': 21590,
            'y-dimension': 33020,
          }),
          dict({
            'x-dimension': 29700,
            'y-dimension': 42000,
          }),
          dict({
            'x-dimension': 27940,
            'y-dimension': 43180,
          }),
          dict({
            'x-dimension': list([
              8890,
              28700,
            ]),
            'y-dimension': list([
              12700,
              43180,
            ]),
          }),
        ]),
        'media-source-supported': list([
          'auto',
          'main',
        ]),
        'media-supported': list([
          'iso_a4_210x297mm',
          'na_letter_8.5x11in',
          'na_executive_7.25x10.5in',
          'iso_a5_148x210mm',
          'iso_a6_105x148mm',
          'na_index-4x6_4x6in',
          'na_index-5x8_5x8in',
          'oe_photo-l_3.5x5in',
          'na_5x7_5x7in',
          'na_number-10_4.125x9.5in',
          'iso_dl_110x220mm',
          'na_monarch_3.875x7.5in',
          'na_legal_8.5x14in',
          'iso_c5_162x229mm',
          'na_foolscap_8.5x13in',
          'iso_a3_297x420mm',
          'na_ledger_11x17in',
          'custom_min_88.9x127mm',
          'custom_max_287x431.8mm',
        ]),
        'media-top-margin-supported': list([
          300,
          0,
          1200,
          2200,
          2200,
        ]),
        'media-type-supported': list([
          'stationery',
          'stationery-inkjet',
          'photographic-glossy',
        ]),
        'multiple-document-jobs-supported': False,
        'multiple-operation-time-out': 150,
        'natural-language-configured': 'de',
        'operations-supported': list([
          <IppOperation.PRINT_JOB: 2>,
          <IppOperation.VALIDATE_JOB: 4>,
          <IppOperation.CREATE_JOB: 5>,
          <IppOperation.SEND_DOCUMENT: 6>,
          <IppOperation.CANCEL_JOB: 8>,
          <IppOperation.GET_JOB_ATTRIBUTES: 9>,
          <IppOperation.GET_JOBS: 10>,
          <IppOperation.GET_PRINTER_ATTRIBUTES: 11>,
          <IppOperation.IDENTIFY_PRINTER: 60>,
        ]),
        'orientation-requested-default': <IppOrientationRequested.PORTRAIT: 3>,
        'orientation-requested-supported': list([
          <IppOrientationRequested.PORTRAIT: 3>,
          <IppOrientationRequested.LANDSCAPE: 4>,
        ]),
        'output-bin-default': 'face-up',
        'output-bin-supported': 'face-up',
        'output-mode-default': 'color',
        'output-mode-supported': list([
          'color',
          'auto',
          'monochrome',
        ]),
        'pages-per-minute': 35,
        'pages-per-minute-color': 27,
        'pdf-versions-supported': 'none',
        'pdl-override-supported': 'attempted',
        'print-color-mode-default': 'color',
        'print-color-mode-supported': list([
          'auto',
          'color',
          'monochrome',
        ]),
        'print-content-optimize-default': 'auto',
        'print-content-optimize-supported': 'auto',
        'print-quality-default': <IppPrintQuality.NORMAL: 4>,
        'print-quality-supported': list([
          <IppPrintQuality.NORMAL: 4>,
          <IppPrintQuality.HIGH: 5>,
        ]),
        'print-scaling-default': 'auto',
        'print-scaling-supported': list([
          'auto',
          'auto-fit',
          'fill',
          'fit',
          'none',
        ]),
        'printer-device-id': 'MFG:Brother;CMD:HBP,BRPJL,URF;MDL:MFC-J5320DW;CLS:PRINTER;CID:Brother Generic Jpeg Type1;URF:SRGB24,W8,CP1,IS1,MT1-8-11,OB9,PQ4-5,RS200-300,OFU0,V1.3,DM4;',
        'printer-dns-sd-name': 'Brother MFC-J5320DW',
        'printer-geo-location': '',
        'printer-icons': list([
          'http://192.168.11.20/ipp/printer-icons-128.png',
          'http://192.168.11.20/ipp/printer-icons-512.png',
        ]),
        'printer-info': 'Brother MFC-J5320DW',
        'printer-is-accepting-jobs': True,
        'printer-kind': list([
          'document',
          'envelope',
          'photo',
        ]),
        'printer-location': '',
        'printer-make-and-model': 'Brother MFC-J5320DW',
        'printer-more-info': 'http://192.168.11.20/net/net/airprint.html',
        'printer-name': 'brother-printer',
        'printer-resolution-default': tuple(
          300,
          300,
          3,
        ),
        'printer-resolution-supported': tuple(
          300,
          300,
          3,
        ),
        'printer-state': <IppPrinterState.IDLE: 3>,
        'printer-state-reasons': 'marker-supply-low-warning',
        'printer-supply-info-uri': 'http://192.168.11.20/general/status.html',
        'printer-up-time': 1326249,
        'printer-uri-supported': 'ipp://192.168.11.20/ipp/print',
        'printer-uuid': 'urn:uuid:e3248000-80ce-11db-8000-30055ce13be2',
        'pwg-raster-document-resolution-supported': tuple(
          300,
          300,
          3,
        ),
        'pwg-raster-document-sheet-back-supported': 'Rotated',
        'pwg-raster-document-type-supported': list([
          'srgb_8',
          'sgray_8',
        ]),
        'queued-job-count': 0,
        'sides-default': 'one-sided',
        'sides-supported': list([
          'one-sided',
          'two-sided-long-edge',
          'two-sided-short-edge',
        ]),
        'urf-supported': list([
          'SRGB24',
          'W8',
          'CP1',
          'IS1',
          'MT1-8-11',
          'OB9',
          'PQ4-5',
          'RS200-300',
          'OFU0',
          'V1.3',
          'DM4',
        ]),
        'uri-authentication-supported': 'none',
        'uri-security-supported': 'none',
      }),
    ]),
    'request-id': 93687,
    'status-code': 0,
    'unsupported-attributes': list([
    ]),
    'version': tuple(
      2,
      0,
    ),
  })
# ---
# name: test_parse_epson_xp6000
  dict({
    'data': b'',
    'jobs': list([
    ]),
    'operation-attributes': dict({
      'attributes-charset': 'utf-8',
      'attributes-natural-language': 'en',
    }),
    'printers': list([
      dict({
        'charset-configured': 'utf-8',
        'charset-supported': 'utf-8',
        'color-supported': True,
        'compression-supported': list([
          'none',
          'gzip',
        ]),
        'copies-default': 1,
        'copies-supported': list([
          1,
          99,
        ]),
        'document-format-default': 'application/octet-stream',
        'document-format-preferred': 'image/urf',
        'document-format-supported': list([
          'application/octet-stream',
          'image/pwg-raster',
          'image/urf',
          'image/jpeg',
        ]),
        'document-format-varying-attributes': list([
          'copies',
          'sides',
        ]),
        'finishings-default': <IppFinishing.NONE: 3>,
        'finishings-supported': <IppFinishing.NONE: 3>,
        'generated-natural-language-supported': 'en',
        'identify-actions-default': 'flash',
        'identify-actions-supported': 'flash',
        'ipp-features-supported': list([
          'wfds-print-1.0',
          'airprint-1.7',
        ]),
        'ipp-versions-supported': list([
          '1.0',
          '1.1',
          '2.0',
        ]),
        'job-creation-attributes-supported': list([
          'copies',
          'finishings',
          'ipp-attribute-fidelity',
          'job-name',
          'media',
          'media-col',
          'orientation-requested',
          'output-bin',
          'print-quality',
          'printer-resolution',
          'sides',
          'print-color-mode',
          'print-content-optimize',
          'print-scaling',
          'job-mandatory-attributes',
        ]),
        'jpeg-features-supported': 'none',
        'jpeg-k-octets-supported': list([
          0,
          16384,
        ]),
        'jpeg-x-dimension-supported': list([
          0,
          15000,
        ]),
        'jpeg-y-dimension-supported': list([
          1,
          15000,
        ]),
        'landscape-orientation-requested-preferred': 5,
        'marker-colors': list([
          '#000000',
          '#00FFFF',
          '#FF00FF',
          '#FFFF00',
          '#000000',
        ]),
        'marker-high-levels': list([
          100,
          100,
          100,
          100,
          100,
        ]),
        'marker-levels': list([
          27,
          99,
          83,
          6,
          64,
        ]),
        'marker-low-levels': list([
          15,
          15,
          15,
          15,
          15,
        ]),
        'marker-names': list([
          'Photo Black ink',
          'Cyan ink',
          'Magenta ink',
          'Yellow ink',
          'Black ink',
        ]),
        'marker-types': list([
          'ink-cartridge',
          'ink-cartridge',
          'ink-cartridge',
          'ink-cartridge',
          'ink-cartridge',
        ]),
        'media-bottom-margin-supported': list([
          0,
          300,
        ]),
        'media-col-default': dict({
          'media-bottom-margin': 300,
          'media-left-margin': 300,
          'media-right-margin': 300,
          'media-size': dict({
            'x-dimension': 21590,
            'y-dimension': 27940,
          }),
          'media-source': 'main',
          'media-top-margin': 300,
          'media-type': 'stationery',
        }),
        'media-col-ready': list([
          dict({
            'media-bottom-margin': 300,
            'media-left-margin': 300,
            'media-right-margin': 300,
            'media-size': dict({
              'x-dimension': 21590,
              'y-dimension': 27940,
            }),
            'media-source': 'main',
            'media-top-margin': 300,
            'media-type': 'stationery',
          }),
          dict({
            'media-bottom-margin': 300,
            'media-left-margin': 300,
            'media-right-margin': 300,
            'media-size': dict({
              'x-dimension': 10160,
              'y-dimension': 15240,
            }),
            'media-source': 'photo',
            'media-top-margin': 300,
            'media-type': 'photographic',
          }),
          dict({
            'media-bottom-margin': 0,
            'media-left-margin': 0,
            'media-right-margin': 0,
            'media-size': dict({
              'x-dimension': 10160,
              'y-dimension': 15240,
            }),
            'media-source': 'photo',
            'media-top-margin': 0,
            'media-type': 'photographic',
          }),
          dict({
            'media-bottom-margin': 0,
            'media-left-margin': 0,
            'media-right-margin': 0,
            'media-size': dict({
              'x-dimension': 12000,
              'y-dimension': 12000,
            }),
            'media-source': 'disc',
            'media-top-margin': 0,
            'media-type': 'disc',
          }),
        ]),
        'media-col-supported': list([
          'media-size',
          'media-top-margin',
          'media-left-margin',
          'media-right-margin',
          'media-bottom-margin',
          'media-type',
          'media-source',
        ]),
        'media-default': 'na_letter_8.5x11in',
        'media-left-margin-supported': list([
          0,
          300,
        ]),
        'media-ready': list([
          'na_letter_8.5x11in',
          'na_index-4x6_4x6in',
          'disc_12cm_18x120mm',
        ]),
        'media-right-margin-supported': list([
          0,
          300,
        ]),
        'media-size-supported': list([
          dict({
            'x-dimension': 21590,
            'y-dimension': 27940,
          }),
          dict({
            'x-dimension': 10160,
            'y-dimension': 15240,
          }),
          dict({
            'x-dimension': 12700,
            'y-dimension': 17780,
          }),
          dict({
            'x-dimension': 20320,
            'y-dimension': 25400,
          }),
          dict({
            'x-dimension': 10160,
            'y-dimension': 18060,
          }),
          dict({
            'x-dimension': 21000,
            'y-dimension': 29700,
          }),
          dict({
            'x-dimension': 10500,
            'y-dimension': 14800,
          }),
          dict({
            'x-dimension': 21590,
            'y-dimension': 35560,
          }),
          dict({
            'x-dimension': 8890,
            'y-dimension': 12700,
          }),
          dict({
            'x-dimension': 13970,
            'y-dimension': 21590,
          }),
          dict({
            'x-dimension': 10477,
            'y-dimension': 24130,
          }),
          dict({
            'x-dimension': 21590,
            'y-dimension': 33020,
          }),
          dict({
            'x-dimension': 12000,
            'y-dimension': 12000,
          }),
          dict({
            'x-dimension': list([
              8900,
              21590,
            ]),
            'y-dimension': list([
              12700,
              111760,
            ]),
          }),
        ]),
        'media-source-supported': list([
          'auto',
          'main',
          'photo',
          'disc',
        ]),
        'media-supported': list([
          'na_letter_8.5x11in',
          'na_index-4x6_4x6in',
          'na_5x7_5x7in',
          'na_govt-letter_8x10in',
          'om_hivision_101.6x180.6mm',
          'iso_a4_210x297mm',
          'iso_a6_105x148mm',
          'na_legal_8.5x14in',
          'oe_photo-l_3.5x5in',
          'na_invoice_5.5x8.5in',
          'na_number-10_4.125x9.5in',
          'na_foolscap_8.5x13in',
          'disc_12cm_18x120mm',
          'custom_min_89x127mm',
          'custom_max_215.9x1117.6mm',
        ]),
        'media-top-margin-supported': list([
          0,
          300,
        ]),
        'media-type-supported': list([
          'stationery',
          'photographic-high-gloss',
          'photographic',
          'photographic-semi-gloss',
          'photographic-glossy',
          'photographic-matte',
          'com.epson-luster',
          'envelope',
          'stationery-coated',
          'disc',
        ]),
        'mopria-certified': 'mopria-certified 1.3',
        'multiple-document-jobs-supported': False,
        'multiple-operation-time-out': 120,
        'multiple-operation-time-out-action': 'abort-job',
        'natural-language-configured': 'en',
        'operations-supported': list([
          <IppOperation.PRINT_JOB: 2>,
          <IppOperation.VALIDATE_JOB: 4>,
          <IppOperation.CREATE_JOB: 5>,
          <IppOperation.SEND_DOCUMENT: 6>,
          <IppOperation.CANCEL_JOB: 8>,
          <IppOperation.GET_JOB_ATTRIBUTES: 9>,
          <IppOperation.GET_JOBS: 10>,
          <IppOperation.GET_PRINTER_ATTRIBUTES: 11>,
          <IppOperation.CLOSE_JOB: 59>,
          <IppOperation.IDENTIFY_PRINTER: 60>,
        ]),
        'orientation-requested-default': <IppOrientationRequested.PORTRAIT: 3>,
        'orientation-requested-supported': <IppOrientationRequested.PORTRAIT: 3>,
        'output-bin-default': 'face-up',
        'output-bin-supported': 'face-up',
        'pages-per-minute': 9,
        'pages-per-minute-color': 9,
        'pdf-versions-supported': 'none',
        'pdl-override-supported': 'attempted',
        'print-color-mode-default': 'auto',
        'print-color-mode-supported': list([
          'color',
          'monochrome',
          'auto-monochrome',
          'process-monochrome',
          'auto',
        ]),
        'print-content-optimize-default': 'auto',
        'print-content-optimize-supported': 'auto',
        'print-quality-default': <IppPrintQuality.NORMAL: 4>,
        'print-quality-supported': list([
          <IppPrintQuality.NORMAL: 4>,
          <IppPrintQuality.HIGH: 5>,
        ]),
        'print-scaling-default': 'auto',
        'print-scaling-supported': list([
          'auto',
          'auto-fit',
          'fill',
          'fit',
          'none',
        ]),
        'printer-alert': 'code=other',
        'printer-alert-description': 'feed roller needed soon',
        'printer-config-change-date-time': '',
        'printer-config-change-time': 25,
        'printer-current-time': datetime.datetime(2022, 10, 4, 2, 21, 58, tzinfo=datetime.timezone.utc),
        'printer-device-id': 'MFG:EPSON;CMD:ESCPL2,BDC,D4,D4PX,ESCPR7,END4,GENEP,URF;MDL:XP-6000 Series;CLS:PRINTER;DES:EPSON XP-6000 Series;CID:EpsonRGB;FID:FXN,DPA,WFA,ETN,AFN,DAN,WRA;RID:20;DDS:022500;ELG:1000;SN:583434593035343012;URF:CP1,PQ4-5,OB9,OFU0,RS360,SRGB24,W8,DM3,IS1-7-6,V1.4,MT1-3-7-8-10-11-12;',
        'printer-dns-sd-name': 'EPSON XP-6000 Series',
        'printer-firmware-name': 'Firmware',
        'printer-firmware-string-version': '20.44.NU25M7',
        'printer-firmware-version': '000020440000M7250000000000000000',
        'printer-geo-location': '',
        'printer-get-attributes-supported': 'document-format',
        'printer-icons': list([
          'https://192.168.1.92:443/PRESENTATION/AIRPRINT/PRINTER_128.PNG',
          'https://192.168.1.92:443/PRESENTATION/AIRPRINT/PRINTER_512.PNG',
        ]),
        'printer-info': 'EPSON XP-6000 Series',
        'printer-input-tray': list([
          'type=other;dimunit=micrometers;mediafeed=279400;mediaxfeed=215900;maxcapacity=-2;level=-2;status=0;name=Sheet feeder bin 1;',
          'type=sheetFeedAutoNonRemovableTray;dimunit=micrometers;mediafeed=279400;mediaxfeed=215900;maxcapacity=-2;level=-2;status=0;name=Sheet feeder bin 1;',
          'type=sheetFeedAutoNonRemovableTray;dimunit=micrometers;mediafeed=152400;mediaxfeed=101600;maxcapacity=-2;level=-2;status=0;name=Sheet feeder bin 2;',
          'type=other;dimunit=micrometers;mediafeed=120000;mediaxfeed=120000;maxcapacity=-2;level=-2;status=5;name=Disc;',
        ]),
        'printer-is-accepting-jobs': True,
        'printer-kind': list([
          'document',
          'envelope',
          'photo',
          'disc',
        ]),
        'printer-location': '',
        'printer-make-and-model': 'EPSON XP-6000 Series',
        'printer-more-info': 'http://192.168.1.92:80/PRESENTATION/BONJOUR',
        'printer-name': 'ipp/print',
        'printer-organization': '',
        'printer-organizational-unit': '',
        'printer-output-tray': 'type=unRemovableBin;maxcapacity=50;remaining=-3;status=0;name=Face-up Tray;stackingorder=lastToFirst;pagedelivery=faceUp;',
        'printer-resolution-default': tuple(
          360,
          360,
          3,
        ),
        'printer-resolution-supported': list([
          tuple(
            360,
            360,
            3,
          ),
          tuple(
            720,
            720,
            3,
          ),
          tuple(
            5760,
            1440,
            3,
          ),
        ]),
        'printer-state': <IppPrinterState.IDLE: 3>,
        'printer-state-change-date-time': datetime.datetime(2022, 9, 27, 3, 47, 19, tzinfo=datetime.timezone.utc),
        'printer-state-change-time': 184119,
        'printer-state-reasons': 'marker-supply-low-warning',
        'printer-strings-languages-supported': list([
          'en',
          'es-mx',
          'pt',
          'fr',
        ]),
        'printer-strings-uri': 'http://192.168.1.92:80/LANGUAGES/IPP?LANG=en',
        'printer-supply-info-uri': 'http://192.168.1.92:80/PRESENTATION/HTML/TOP/PRTINFO.HTML',
        'printer-up-time': 783801,
        'printer-uri-supported': list([
          'ipps://192.168.1.92:631/ipp/print',
          'ipp://192.168.1.92:631/ipp/print',
        ]),
        'printer-uuid': 'urn:uuid:cfe92100-67c4-11d4-a45f-f8d027761251',
        'pwg-raster-document-resolution-supported': tuple(
          360,
          360,
          3,
        ),
        'pwg-raster-document-sheet-back': 'rotated',
        'pwg-raster-document-type-supported': list([
          'sgray_8',
          'srgb_8',
        ]),
        'queued-job-count': 0,
        'sides-default': 'one-sided',
        'sides-supported': list([
          'one-sided',
          'two-sided-short-edge',
          'two-sided-long-edge',
        ]),
        'urf-supported': list([
          'CP1',
          'PQ4-5',
          'OB9',
          'OFU0',
          'RS360',
          'SRGB24',
          'W8',
          'DM3',
          'IS1-7-6',
          'V1.4',
          'MT1-3-7-8-10-11-12',
        ]),
        'uri-authentication-supported': list([
          'none',
          'none',
        ]),
        'uri-security-supported': list([
          'tls',
          'none',
        ]),
        'which-jobs-supported': list([
          'completed',
          'not-completed',
        ]),
      }),
    ]),
    'request-id': 66306,
    'status-code': 0,
    'unsupported-attributes': list([
    ]),
    'version': tuple(
      2,
      0,
    ),
  })
# ---
# name: test_parse_kyocera_ecosys_m2540dn
  dict({
    'data': b'',
    'jobs': list([
    ]),
    'operation-attributes': dict({
      'attributes-charset': 'utf-8',
      'attributes-natural-language': 'en-us',
    }),
    'printers': list([
      dict({
        'printer-info': 'mfu00-0365',
        'printer-location': '8409',
        'printer-make-and-model': 'ECOSYS M2540dn',
        'printer-name': 'mfu00-0365',
        'printer-state': <IppPrinterState.IDLE: 3>,
        'printer-state-message': 'Sleeping...  ',
        'printer-uri-supported': list([
          'ipps://10.104.12.95:443/ipp/print',
          'ipp://10.104.12.95:631/ipp/print',
        ]),
      }),
    ]),
    'request-id': 47131,
    'status-code': 1,
    'unsupported-attributes': list([
      dict({
        'requested-attributes': list([
          'printer-type',
          'printer-state-reason',
          'device-uri',
          'printer-is-shared',
        ]),
      }),
    ]),
    'version': tuple(
      2,
      0,
    ),
  })
# ---
